SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -46,15 +46,43 @@ void USpatialLoadBalanceEnforcer::OnAuthorityIntentComponentUpdated(const Worker
 	}
 }
 
+// This is called whenever this worker becomes authoritative for the ACL component on an entity.
+// It is now this worker's responsibility to make sure that the ACL reflects the current intent,
+// which may have been received before this call.
 void USpatialLoadBalanceEnforcer::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
 {
 	if (AuthOp.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID &&
 		AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE)
 	{
+		const AuthorityIntent* AuthorityIntentComponent = StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(AuthOp.entity_id);
+		if (AuthorityIntentComponent == nullptr)
+		{
+			// TODO(zoning): There are still some entities being created without an authority intent component.
+			// For example, the Unreal created worker entities don't have one. Even though those won't be able to
+			// transition, we should have the intent component on them for completeness.
+		 	UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Requested authority change for entity without AuthorityIntent component. EntityId: %lld"), AuthOp.entity_id);
+			return;
+		}
+
+		const FString* OwningWorkerId = VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
+		if (OwningWorkerId != nullptr &&
+			*OwningWorkerId == WorkerId &&
+			StaticComponentView->GetAuthority(AuthOp.entity_id, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID) == WORKER_AUTHORITY_AUTHORITATIVE)
+		{
+			UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose, TEXT("No need to queue newly authoritative entity %lld because this worker is already authoritative."), AuthOp.entity_id);
+			return;
+		}
 		QueueAclAssignmentRequest(AuthOp.entity_id);
 	}
 }
 
+// QueueAclAssignmentRequest is called from three places.
+// 1) AuthorityIntent change - Intent is not authoritative on this worker - ACL is authoritative on this worker.
+//    (another worker changed the intent, but this worker is responsible for the ACL, so update it.)
+// 2) ACL change - Intent is unknown - ACL just became authoritative on this worker.
+//    (this worker just became responsible, so check to make sure intent and ACL agree.)
+// 3) AuthorityIntent change - Intent is authoritative on this worker but no longer assigned to this worker - ACL is authoritative on this worker.
+//    (this worker had responsibility for both and is giving up authority.)
 void USpatialLoadBalanceEnforcer::QueueAclAssignmentRequest(const Worker_EntityId EntityId)
 {
 	if (AclWriteAuthAssignmentRequests.ContainsByPredicate([EntityId](const WriteAuthAssignmentRequest& Request) { return Request.EntityId == EntityId; }))
```

Comment:
How frequent is this check and how large is the `AclWriteAuthAssignmentRequests` array? Should we use a map instead?

Reply:
So this is added to when an actors cross a zone boundary. This probably probably means the container is going to be quite small.  That said, depending on your game perhaps it gets a little large? Perhaps we should have a simple way to expose how large it gets so we can monitor it. I think the issue with a map might be ordering. I think when we process this data, we might want to iterate through it in the order that things were added to it. If order is is not important and the size of the contain frequently gets "large" then maybe a map is the way to go.

Reply:
If in doubt use an array. You iterate over this on the same order as add from it and then remove everything basically in one go. I suspect you need a fair few requests being made before it's better to use a map.

Reply:
I added a TODO.

Reply:
Right, since we need to preserve the ordering, I agree we should use an array. Resolving.